wandb/sdk/wandb_settings.py

```diff
@@ -223,925 +146,1015 @@ class SettingsConsole(enum.Enum):
+            "value": "True",
+            "preprocessor": _str_as_bool,
+            "validator": lambda x: isinstance(x, bool),
+        }
+        self.show_warnings: Any = {
+            "value": "True",
+            "preprocessor": _str_as_bool,
+            "validator": lambda x: isinstance(x, bool),
+        }
+        self.silent: Any = {
+            "value": "False",
+            "preprocessor": _str_as_bool,
+            "validator": lambda x: isinstance(x, bool),
+        }
+        self.start_method: Any = {
+            "validator": [lambda x: isinstance(x, str), self._validate_start_method],
+        }
+        self.strict: Any = {
+            "preprocessor": _str_as_bool,
+            "validator": lambda x: isinstance(x, bool),
+        }
+        self.summary_errors: Any = {
+            "validator": lambda x: isinstance(x, int),
+        }
+        self.summary_warnings: Any = {
+            "value": 5,
+            "preprocessor": lambda x: int(x),
+            "validator": lambda x: isinstance(x, int),
+            "is_policy": True,
+        }
+        self.sweep_id: Any = {
+            "validator": lambda x: isinstance(x, str),
+        }
+        self.sweep_param_path: Any = {
+            "validator": lambda x: isinstance(x, str),
+        }
+        self.symlink: Any = {
+            "preprocessor": _str_as_bool,
+            "validator": lambda x: isinstance(x, bool),  # probed
+        }
+        self.sync_dir: Any = {
+            "value": "<sync_dir>",
+            "validator": lambda x: isinstance(x, str),
+            "hook": [
+                lambda x: self._path_convert(
+                    self.wandb_dir, f"{self.run_mode}-{self.timespec}-{self.run_id}"
+                )
+            ],
+        }
+        self.sync_file: Any = {
+            "value": "run-<run_id>.wandb",
+            "validator": lambda x: isinstance(x, str),
+            "hook": lambda x: self._path_convert(
+                self.sync_dir, f"run-{self.run_id}.wandb"
+            ),
+        }
+        self.sync_symlink_latest: Any = {
+            "value": "latest-run",
+            "validator": lambda x: isinstance(x, str),
+            "hook": lambda x: self._path_convert(self.wandb_dir, x),
+        }
+        self.system_sample: Any = {
+            "value": 15,
+            "validator": lambda x: isinstance(x, int),
+        }
+        self.system_sample_seconds: Any = {
+            "value": 2,
+            "validator": lambda x: isinstance(x, int),
+        }
+        self.tmp_dir: Any = {
+            "value": "tmp",
+            "validator": lambda x: isinstance(x, str),
+            "hook": lambda x: (
+                self._path_convert(
+                    self.wandb_dir, f"{self.run_mode}-{self.timespec}-{self.run_id}", x
+                )
+                or tempfile.gettempdir()
+            ),
+        }
+        self.username: Any = {
+            "validator": lambda x: isinstance(x, str),
+        }
```

Comment:
All of these `Any` annotations make me feel sad.  Any reason not to make these NamedTuples or something with strong type gurarentees?

Reply:
All these are converted to `Property` class instances immediately after definition, all of which come with validators that guarantee not only types, but also anything else, specific to each setting (see e.g. https://github.com/wandb/client/pull/3083/files/cf4a6864f7e306dd5d3fc1172fb26f29fc96265f#diff-432d500a1d77db4e4a22825d98e99cb860fee78777f6540b85395c1e5ee28e94R476). Here's my reasoning for doing it this way:
- Properties encapsulate all the logic we want from them: preprocessing, validation, runtime hooks, deps, source priorities (see an example in the pr description). Yes, static checkers aren't of much use in this case, action happens at runtime. But we should bring our test coverage to ~100% anyways :) 
  - Validation allows programming by contract of sorts (long live [PEP316](https://www.python.org/dev/peps/pep-0316/) ^_^), which static checkers won't help with.
- I could have gone with `object.__setattr__` to define things (like we're doing now) but that breaks code completion in pycharm (and i'd expect same in vscode), and it's kind of nice to have.
- We could define these as Properties straight away:
```
self.mofl: Property = Property(
     name="mofl",
     validator=[lambda x: isinstance(x, (int, float)), lambda x: x > 42],
 )
```
or something, it just takes up more space to spell it out since I decided to attach names to them. I'm afraid it might freak mypy out, although I haven't tried to be honest.

Reply:
@raubitsj has a great idea about how to get the best of both static and runtime worlds, implementing now.
[the idea is to use typed class attributes + dynamic re-definition as properties propagating type annotation as the first validator in the resulting property object; this does add a bit of pain to the dev experience, but I'm adding a note to the contrib guide + a unit test that ensures people don't screw up in the process].